### PR TITLE
[BugFix] Fix use of per-request seed with pipeline parallel

### LINKED ISSUE #6449: [Bug]: Seed issue with Pipeline Parallel
### Your current environment

v0.5.1

### 🐛 Describe the bug

OpenAI API specifies that you can provide a seed: https://platform.openai.com/docs/api-reference/chat/create#chat-create-seed

This allows reproducibility for example with non-zero temperature parameter.

Currently, any state information is stored/advanced on the driver process only. We need to extend this to the worker actually doing the sampling.

### PR DESCRIPTION
**Pull request by njhill**
The current per-request seed implementation assumes that the sampling happens in the same process as the driver worker since the sampler `SequenceGroup` objects are used to hang the `torch.Generators` on to maintain state between steps.

This assumption is now not always true and in particular this is why seeds are currently broken with pipeline parallel.

This PR moves the per-sequence group generator state to a dict in the final-rank PP worker where the sampler resides. Now that `finished_requests_ids` are passed in the `execute_model` calls they can be used to clean out the state for completed requests.

For speculative decoding, the generators from the scorer model worker are used by the rejection sampler.

Changes are also included to simplify/optimize the seeded spec decoding paths, and a seeded mlp speculator CI test is added.

Fixes https://github.com/vllm-project/vllm/issues/6449

### CODE REVIEW

**Comment by njhill** [comment]
Draft because I'm still making sure this works properly with spec decoding.

File: vllm/worker/model_runner.py [hunk 1]
```diff
@@ -1264,11 +1264,15 @@ def prepare_model_input(
         """
         model_input = self._prepare_model_input_tensors(
             seq_group_metadata_list, finished_requests_ids)
-        sampling_metadata = SamplingMetadata.prepare(seq_group_metadata_list,
-                                                     model_input.seq_lens,
-                                                     model_input.query_lens,
-                                                     self.device,
-                                                     self.pin_memory)
+        if get_pp_group().is_last_rank:
```
Comment (njhill) [inline vllm/worker/model_runner.py:1267]: @andoorve I noticed this small optimization - only need to prepare the sampling metadata tensors in the last rank

Comment (andoorve) [reply vllm/worker/model_runner.py:1267]: Thanks! There's a lot of these optimizations we didn't include last time. Just from what I can tell, I think this one in particular might not help E2E (will just increase the bubble size on the other workers) but doesn't hurt to have

Comment (njhill) [reply vllm/worker/model_runner.py:1267]: Yeah I figured that but doesn't harm to skip redundant work I guess

Comment (andoorve) [reply vllm/worker/model_runner.py:1267]: Yup for sure

**Comment by andoorve** [comment]
Will take a look today!

File: vllm/worker/model_runner_base.py [hunk 2]
```diff
@@ -176,3 +179,15 @@ def execute_model(
...
         """
         raise NotImplementedError
+
+    def get_generators(self, finished_request_ids: Optional[List[str]] = None):
+        """
+        Return dict of per-request generators used for random sampling.
+        """
+
+        # Clean up generators from completed requests
+        if finished_request_ids:
```
Comment (andoorve) [inline vllm/worker/model_runner_base.py:189]: Could it be possible that we run into the same abort issue we had previously? I.e. could we abort, get rid of a generator on another "virtual engine" and then find that we're missing the right generator when we get to the part where we use it?

**Review by andoorve — APPROVED** [review]
As long as the above is satisfied I don't have an issue it from my POV. Also would be good to check that it works on the below settings from @aurickq which previously did not work with PP:

Can be triggered using PP=2 and OpenAI Server with the following repro:
```
kwargs = {
    'model': 'meta-llama/Meta-Llama-3-70B',
    'prompt': [
        [14924, 25, 14693, 39710, 374, 264],
        [14924, 25, 14693, 39710, 374, 264],
    ],
    'echo': True,
    'max_tokens': 0,
    'temperature': 0.0,
    'logprobs': 1,
    'seed': 1234,
}

completion = client.completions.create(**kwargs)
```

Other than that would recommend someone to look at the spec decoding part.

File: vllm/worker/model_runner_base.py (hunk 2 above)
Comment (njhill) [reply vllm/worker/model_runner_base.py:189]: @andoorve I think this is fine since the dict storing the generators is indexed by request id and shared between all of the virtual engines. In fact in this case the `finished_request_ids` come from the VE-specific scheduler so the VE tasks will only be cleaning up their own ones, but even if that wasn't the case it would still work fine.

**Comment by njhill** [comment]
Thanks @andoorve! I should also point out that this PP seeded generation is now tested in the CI, with [this](https://github.com/vllm-project/vllm/pull/6698/files#diff-9e08e8c7769db7b58089fd7659d75ca96df40bb7b5d50299d3a30abd54da9ad6R182) addition to the `compare_two_settings` method used by `test_pipeline_parallel.py`.

**Comment by andoorve** [comment]
> Thanks @andoorve! I should also point out that this PP seeded generation is now tested in the CI, with [this](https://github.com/vllm-project/vllm/pull/6698/files#diff-9e08e8c7769db7b58089fd7659d75ca96df40bb7b5d50299d3a30abd54da9ad6R182) addition to the compare_two_settings method used by test_pipeline_parallel.py.

Oh thanks for pointing that out! I missed it. Should we add a test with batched prompts as well with seed? Apologies if I missed it. I think the condition that @aurickq triggered only happens with multiple prompt.

**Comment by njhill** [comment]
Thanks @andoorve, I've now added a comparison test with seed and multiple prompts.

**Comment by njhill** [comment]
@tdoublep would you mind looking over the spec decoding related changes?

**Review by youkaichao — APPROVED** [review]
I gave you approval to unblock you. Please feel free to merge when you think it is ready @njhill @andoorve

File: vllm/spec_decode/batch_expansion.py [hunk 3]
```diff
@@ -247,24 +249,36 @@ def _create_target_seq_group_metadata(
         token_ids_to_score = self._get_token_ids_to_score(
             proposal_token_ids[batch_index])
 
+        # Use simpler sampling parameters apart from for final token
+        # (in particular don't do seeded sampling) since those sampled tokens
+        # aren't used
+        sampling_params = input_seq_group_metadata.sampling_params
+        non_bonus_sampling_params = DEFAULT_SIMPLE_SAMPLING_PARAMS \
+            if sampling_params.temperature else sampling_params
```
Comment (tdoublep) [inline vllm/spec_decode/batch_expansion.py:257]: Can you help me understand why we should use a default sampling params if the temperature is *not* None?

File: vllm/spec_decode/batch_expansion.py [hunk 4]
```diff
@@ -247,24 +249,36 @@ def _create_target_seq_group_metadata(
         token_ids_to_score = self._get_token_ids_to_score(
             proposal_token_ids[batch_index])
 
+        # Use simpler sampling parameters apart from for final token
+        # (in particular don't do seeded sampling) since those sampled tokens
+        # aren't used
```
Comment (tdoublep) [inline vllm/spec_decode/batch_expansion.py:254]: I don't completely follow this optimization. Do we really only use the sampled token in the case that all proposals are accepted? My understanding was that if we accept e.g., 2 proposed tokens, we can still get a third one by sampling after verification? Don't we need the sampling params to be correct in all cases?

Comment (njhill) [reply vllm/spec_decode/batch_expansion.py:254]: > Do we really only use the sampled token in the case that all proposals are accepted?

Yes, it's only used for the bonus token.

> My understanding was that if we accept e.g., 2 proposed tokens, we can still get a third one by sampling after verification?

Yes but in this case the rejection sampler uses a "recovered" distribution based on the target and draft logits, so it's essentially a separate sampling operation. Sampled tokens from the batch-expanded target forward aren't used.

This can be seen [here](https://github.com/vllm-project/vllm/blob/main/vllm/spec_decode/spec_decode_worker.py#L585) where just the final sampled token is sliced off from the contracted sequence.

Comment (tdoublep) [reply vllm/spec_decode/batch_expansion.py:254]: Right, I see now. The rejection sampler uses the recovered distribution if we reject one of the proposals, and indeed the only case where this doesn't happen is if *all* proposals are accepted (e.g. bonus token logic).

File: vllm/spec_decode/batch_expansion.py (hunk 3 above)
Comment (tdoublep) [reply vllm/spec_decode/batch_expansion.py:257]: I see, thx. I actually was digging into that part of the code the other day to understand how the rejection sampler worked in the greedy case, it makes sense. My confusion mainly stemmed from thinking temperature was optional.

**Review by tdoublep — APPROVED** [review]
LGTM

**Comment by njhill** [comment]
Thanks @tdoublep @andoorve!
